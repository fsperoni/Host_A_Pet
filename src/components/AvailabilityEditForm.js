import React, { useState, useContext } from "react";
import Alert from "./Alert";
import HostAPetApi from "./Api";
import UserContext from "../hooks/useUserContext";

/** Edit Availability form. */

const AvailabilityEditForm = ({avail, avails, setAvails, setShowEditForm, roles}) => {
  const min = new Date().toISOString().substring(0,10);
  const { currentUser } = useContext(UserContext);
  const [formData, setFormData] = useState({
    startDate: avail.startDate.substring(0,10),
    endDate: avail.endDate.substring(0,10),
    roleId: avail.roleId
  });
  const [formErrors, setFormErrors] = useState([]);
  

  /** Handle form submit */
  const handleSubmit = async (evt) => {
    evt.preventDefault();
    const id = parseInt(evt.target.children[0].value);
    try {
      const result = await HostAPetApi.updateAvailability(currentUser.username, id, formData)
      const newAvails = avails.filter(avail => avail.id !== id);
      setAvails([...newAvails, result])
      setFormErrors([]);
      setShowEditForm(false);
    } catch (err) {
      setFormErrors(err);
    }
  }

  /** Handle form field changes */
  const handleChange = (evt) => {
    const { name, value } = evt.target;
    setFormData(data => ({ ...data, [name]: value }));
    setFormErrors([]);
  }

  /** Handle form cancel button */
  const handleCancel = (evt) => {
    evt.preventDefault();
    setFormErrors([]);
    setShowEditForm(false);
  }

  const generateOptions = () => {
    return roles.filter(r => r.id !== formData.roleId).map(role => (
      <option key={role.id} value={role.id}>{role.name}</option>
    ))
  }
  const options = generateOptions();
  const role = roles.find(role => role.id === avail.roleId);
  return (
    <div className="AvailabilityEditForm mt-3">
      <div className="container">
        <h2 className="mb-3">Update availability</h2>
        <div className="card">
          <div className="card-body">
            <form onSubmit={handleSubmit}>
              <div className="form-group">
                <label>Start date:</label>
                <input name="startDate" className="form-control" onChange={handleChange}
                  value={formData.startDate} type="date" min={min}
                />
              </div>
              <div className="form-group">
                <label>End date:</label>
                <input name="endDate" className="form-control" onChange={handleChange}
                  value={formData.endDate} type="date" min={min}
                />
              </div>
              <div className="form-group">
                <label>Role:</label>
                <select name="roleId" className="form-select" onChange={handleChange}>
                  <option value={role.id}>{role.name}</option>
                  {options}
                </select>
              </div>
              {formErrors.length>0 && <Alert type="danger" messages={formErrors} />}
              <div className="container mt-4">
              <button 
                className="btn btn-primary btn-block"
                onClick={handleSubmit}
              > <data value={avail.id} />
                Save
              </button>
              <button
                className="btn btn-warning btn-block ms-4"
                onClick={handleCancel}
              >
                Cancel
              </button>
            </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  );
}

export default AvailabilityEditForm;