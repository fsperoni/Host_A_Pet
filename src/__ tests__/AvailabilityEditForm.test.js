import { render } from "@testing-library/react";
import AvailabilityEditForm from "../components/AvailabilityEditForm";
import UserContext from "../hooks/useUserContext";

const {
  currentAdminUser,
  avails, roles, avail
} = require("./_testCommon");

// smoke test
it('renders without crashing', function() {
  render(
    <UserContext.Provider
          value={{ currentUser: currentAdminUser() , function() {}}}>
      <AvailabilityEditForm avail={avail()} avails={avails()} 
        setAvails={function() {}} setShowEditForm={function() {}} roles={roles()}/>
    </UserContext.Provider>);
});

//snapshot test
it("matches snapshot", function() {
  const {asFragment} = render(
    <UserContext.Provider
          value={{ currentUser: currentAdminUser() , function() {}}}>
      <AvailabilityEditForm avail={avail()} avails={avails()} 
        setAvails={function() {}} setShowEditForm={function() {}} roles={roles()}/>
    </UserContext.Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("should have 'Update availability' in the component.", function() {
  const { getByText } = render(
    <UserContext.Provider
          value={{ currentUser: currentAdminUser() , function() {}}}>
      <AvailabilityEditForm avail={avail()} avails={avails()} 
        setAvails={function() {}} setShowEditForm={function() {}} roles={roles()}/>
    </UserContext.Provider>
  );
  expect(getByText("Update availability")).toBeInTheDocument();
});