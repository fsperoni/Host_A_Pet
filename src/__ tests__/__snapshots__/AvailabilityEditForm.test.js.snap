// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="AvailabilityEditForm mt-3"
  >
    <div
      class="container"
    >
      <h2
        class="mb-3"
      >
        Update availability
      </h2>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <form>
            <div
              class="form-group"
            >
              <label>
                Start date:
              </label>
              <input
                class="form-control"
                min="2022-04-02"
                name="startDate"
                type="date"
                value="2022-04-02"
              />
            </div>
            <div
              class="form-group"
            >
              <label>
                End date:
              </label>
              <input
                class="form-control"
                min="2022-04-02"
                name="endDate"
                type="date"
                value="2022-04-20"
              />
            </div>
            <div
              class="form-group"
            >
              <label>
                Role:
              </label>
              <select
                class="form-select"
                name="roleId"
              >
                <option
                  value="1"
                >
                  Host
                </option>
                <option
                  value="2"
                >
                  Pet Owner
                </option>
              </select>
            </div>
            <div
              class="container mt-4"
            >
              <button
                class="btn btn-primary btn-block"
              >
                 
                <data
                  value="1"
                />
                Save
              </button>
              <button
                class="btn btn-warning btn-block ms-4"
              >
                Cancel
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
